applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -676,6 +410,347 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
         }
     }
 
+
+    void InitializeHROMWeights(ModelPart& rModelPart)
+    {
+        KRATOS_TRY
+
+        using ElementQueue = moodycamel::ConcurrentQueue<Element::Pointer>;
+        using ConditionQueue = moodycamel::ConcurrentQueue<Condition::Pointer>;
+
+        // Inspecting elements
+        ElementQueue element_queue;
+        DofsVectorType tmp_dof_list; // Preallocation
+        block_for_each(rModelPart.Elements().GetContainer(), tmp_dof_list,
+            [&](Element::Pointer p_element, DofsVectorType& dof_list)
+        {
+            if (p_element->Has(HROM_WEIGHT)) {
+                element_queue.enqueue(std::move(p_element));
+            } else {
+                p_element->SetValue(HROM_WEIGHT, 1.0);
+            }
+        });
+
+        // Inspecting conditions
+        ConditionQueue condition_queue;
+        block_for_each(rModelPart.Conditions().GetContainer(), tmp_dof_list,
+            [&](Condition::Pointer p_condition, DofsVectorType& dof_list)
+        {
+            if (p_condition->Has(HROM_WEIGHT)) {
+                condition_queue.enqueue(std::move(p_condition));
+            } else {
+                p_condition->SetValue(HROM_WEIGHT, 1.0);
+            }
+        });
+
+        // Dequeueing elements
+        std::size_t err_id;
+        mSelectedElements.reserve(element_queue.size_approx());
+        Element::Pointer p_element;
+        while ( (err_id = element_queue.try_dequeue(p_element)) != 0) {
+            mSelectedElements.push_back(std::move(p_element));
+        }
+
+        // Dequeueing conditions
+        mSelectedConditions.reserve(condition_queue.size_approx());
+        Condition::Pointer p_condition;
+        while ( (err_id = condition_queue.try_dequeue(p_condition)) != 0) {
+            mSelectedConditions.push_back(std::move(p_condition));
+        }
+
+        // Wrap-up
+        mHromSimulation = !(mSelectedElements.empty() && mSelectedConditions.empty());
+        mHromWeightsInitialized = true;
+
+        rModelPart.GetCommunicator().GetDataCommunicator().OrReduceAll(mHromSimulation);
+
+        KRATOS_CATCH("")
+    }
+    
+    static DofQueue ExtractDofSet(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart)
+    {
+        KRATOS_TRY
+
+        DofQueue dof_queue;
+
+        // Emulates ConcurrentQueue::enqueue_bulk adding move semantics to avoid atomic ops
+        const auto enqueue_bulk_move = [](DofQueue& queue, DofsVectorType& dof_list) {
+            for(auto& p_dof: dof_list) {
+                queue.enqueue(std::move(p_dof));
+            }
+            dof_list.clear();
+        };

```

Comment:
```suggestion
        const auto enqueue_bulk_move = [](DofQueue& r_queue, DofsVectorType& r_dof_list) {
            for(auto& p_dof: r_dof_list) {
                r_queue.enqueue(std::move(p_dof));
            }
            r_dof_list.clear();
        };
```
Minor style suggestion.